x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/Contains.java

```diff
@@ -0,0 +1,135 @@
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.ParamOrdinal.FIRST;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.ParamOrdinal.SECOND;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.isString;
+
+/**
+ * Contains function, given a string 'a' and a substring 'b', returns true if the substring 'b' is in 'a'.
+ */
+public class Contains extends EsqlScalarFunction implements OptionalArgument {
+    public static final NamedWriteableRegistry.Entry ENTRY = new NamedWriteableRegistry.Entry(Expression.class, "Contains", Contains::new);
+
+    private final Expression str;
+    private final Expression substr;
+
+    @FunctionInfo(returnType = "boolean", description = """
+        Returns true if a keyword substring is within another string.
+        Returns false if the substring cannot be found.""", examples = @Example(file = "string", tag = "contains"))
+    public Contains(
+        Source source,
+        @Param(name = "string", type = { "keyword", "text" }, description = "An input string") Expression str,
+        @Param(name = "substring", type = { "keyword", "text" }, description = "A substring to find in the input string") Expression substr
+    ) {
+        super(source, Arrays.asList(str, substr));
+        this.str = str;
+        this.substr = substr;
+    }
+
+    private Contains(StreamInput in) throws IOException {
+        this(Source.readFrom((PlanStreamInput) in), in.readNamedWriteable(Expression.class), in.readNamedWriteable(Expression.class));
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        source().writeTo(out);
+        out.writeNamedWriteable(str);
+        out.writeNamedWriteable(substr);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Override
+    public DataType dataType() {
+        return DataType.BOOLEAN;
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+
+        TypeResolution resolution = isString(str, sourceText(), FIRST);
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+        resolution = isString(substr, sourceText(), SECOND);
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+
+        return TypeResolution.TYPE_RESOLVED;
+    }
+
+    @Override
+    public boolean foldable() {
+        return str.foldable() && substr.foldable();
+    }
+
+    @Evaluator
+    static boolean process(BytesRef str, BytesRef substr) {
+        if (str == null || substr == null || str.length < substr.length) {
+            return false;
+        }
+        String utf8ToString = str.utf8ToString();
+        return utf8ToString.contains(substr.utf8ToString());
+    }
```

Comment:
I *think* it's probably faster to do it by hand without converting to a `String` first. But this is fine for now.